### PR TITLE
Depth to Image Generation

### PR DESCRIPTION
**Pull request by pi314ever**
# What does this PR do?

This PR enables Depth to Image generation for [stabilityai/stable-diffusion-2-depth](https://huggingface.co/stabilityai/stable-diffusion-2-depth). Changes were made to the pipeline and DDIM scheduler. 

## Benchmarks

Benchmark was performed between A100 and Gaudi 2 with the following settings:
- `torch_dtype = torch.bfloat16`
- `prompt = "two tigers"`
- `n_prompt = "bad, deformed, ugly, bad anotomy"`
- Initial image: http://images.cocodataset.org/val2017/000000039769.jpg
- Pipeline command: `pipe(prompt=prompt, image=init_image, negative_prompt=n_prompt, num_inference_steps=50)`
- 2 Warmup iterations followed by 10 iterations timed
- Times reported in seconds

Results:
| Batch size | A100 | Gaudi2 /w PNDMScheduler | Gaudi2 /w DDIMScheduler |
| ---- | ---- | ---- | ---- |
| 1 | 1.290 | 1.757 | 1.760 |
| 2 | 2.017 | 2.934 | 2.908 |
| 4 | 3.375 | 5.195 | 5.153 |
| 8 | 6.268 | 9.664 | 9.649 |
| 16 | 11.922 | 18.366 | 18.320 |
| 32 | 23.303 |  35.974 | 34.727 |

> [!Note]
> PNDMScheduler has a much higher latency than DDIMScheduler during warmup

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: tests/test_diffusers.py [hunk 1]
```diff
@@ -1759,6 +1766,238 @@ def test_stable_diffusion_multicontrolnet_hpu_graphs(self):
...
+            output_type="np",
+        )
+        end_time = time.time()
+        latency = end_time - start_time
+        images = outputs.images
+
+        assert len(images) == 1
+        assert images[0].shape == (256, 256, 3)
+
+        assert latency < 1.05 * DEPTH2IMG_GENERATION_LATENCY_BASELINE_BF16
```
Comment (ssarkar2) [inline tests/test_diffusers.py:1998]: Is there a way to test the fidelity of the image to the prompt?

something like this?
https://github.com/huggingface/optimum-habana/pull/920/files#diff-5d657a955747ad953e9a153bda1b8fd7324322d2e19c429c15e749cd9ed39f31

Comment (pi314ever) [reply tests/test_diffusers.py:1998]: The pipeline depends on an input image. Currently, it just draws random numbers for pixel values as a mock input. Are we allowed to add downloads for images? I don't think it is possible to test fidelity otherwise.

Comment (ssarkar2) [reply tests/test_diffusers.py:1998]: I see. Not sure I can remember any other test that has an inp image, so not sure how that might work.

File: examples/stable-diffusion/depth_to_image_generation.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+try:
+    from optimum.habana.utils import check_optimum_habana_min_version
+except ImportError:
+
+    def check_optimum_habana_min_version(*a, **b):
+        return ()
+
+
+# Will error if the minimal version of Optimum Habana is not installed. Remove at your own risks.
+check_optimum_habana_min_version("1.11.0")
```
Comment (regisss) [inline examples/stable-diffusion/depth_to_image_generation.py:42]: Can you rebase and see if it works with main?
```suggestion
check_optimum_habana_min_version("1.14.0.dev0")
```

**Review by regisss — COMMENTED** [review]
I left a few minor comments. Can you rebase this branch on main to make sure the CI works please?

File: optimum/habana/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -179,7 +179,7 @@ def roll_params(self):
     #     return sample
 
     def _get_variance(self, alpha_prod_t, alpha_prod_t_prev):
-        beta_prod_t = 1 - alpha_prod_t
+        beta_prod_t = 1 - alpha_prod_t + 1e-8  # For numerical stability
```
[the remark this answers is not in the document]
Comment (pi314ever) [reply optimum/habana/diffusers/schedulers/scheduling_ddim.py:182]: I'm not sure, I will investigate this with other pipelines.

File: examples/stable-diffusion/README.md [hunk 4]
```diff
@@ -520,6 +520,21 @@ python image_to_image_generation.py \
...
 
+### Depth to Image Generation
+
+Here is how to generate a depth2img-guided image generation using HPU graphs with BF16:
+
+```bash
+python depth_to_image_generation.py \
+    --model_name_or_path "stabilityai/stable-diffusion-2-depth" \
+    --prompts "two tigers" \
+    --base_image "img.png" \
```
Comment (regisss) [inline examples/stable-diffusion/README.md:531]: This is a local path, is it possible to have a path pointing to a hub or github address?

File: tests/test_diffusers.py (hunk 1 above)
Comment (regisss) [reply tests/test_diffusers.py:2232]: Yes, you can download an image as part of the test. Here is an example: https://github.com/huggingface/optimum-habana/blob/6b2a5f843c963a9419ad246c77c3a39d671da198/tests/test_image_classification.py#L46

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+    encoder_output: torch.Tensor, generator: Optional[torch.Generator] = None, sample_mode: str = "sample"
+):
+    if hasattr(encoder_output, "latent_dist") and sample_mode == "sample":
+        return encoder_output.latent_dist.sample(generator)
+    elif hasattr(encoder_output, "latent_dist") and sample_mode == "argmax":
+        return encoder_output.latent_dist.mode()
+    elif hasattr(encoder_output, "latents"):
+        return encoder_output.latents
+    else:
+        raise AttributeError("Could not access latents of provided encoder_output")
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:99]: IT's exactly the same as https://github.com/huggingface/diffusers/blob/v0.29.2/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py#L40 no? Can't we directly import it from there?

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+
+        image = [np.array(i.resize((w, h), resample=PIL_INTERPOLATION["lanczos"]))[None, :] for i in image]
+        image = np.concatenate(image, axis=0)
+        image = np.array(image).astype(np.float32) / 255.0
+        image = image.transpose(0, 3, 1, 2)
+        image = 2.0 * image - 1.0
+        image = torch.from_numpy(image)
+    elif isinstance(image[0], torch.Tensor):
+        image = torch.cat(image, dim=0)
+    return image
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:123]: Same

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+            Variational Auto-Encoder (VAE) model to encode and decode images to and from latent representations.
+        text_encoder ([`~transformers.CLIPTextModel`]):
+            Frozen text-encoder ([clip-vit-large-patch14](https://huggingface.co/openai/clip-vit-large-patch14)).
+        tokenizer ([`~transformers.CLIPTokenizer`]):
+            A `CLIPTokenizer` to tokenize text.
+        unet ([`UNet2DConditionModel`]):
+            A `UNet2DConditionModel` to denoise the encoded image latents.
+        scheduler ([`SchedulerMixin`]):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents. Can be one of
+            [`DDIMScheduler`], [`LMSDiscreteScheduler`], or [`PNDMScheduler`].
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:146]: We can remove this part as it is easily accessible from the link right above.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+    """
+
+    model_cpu_offload_seq = "text_encoder->unet->vae"
+    _callback_tensor_inputs = ["latents", "prompt_embeds", "negative_prompt_embeds", "depth_mask"]
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:150]: We can remove these two lines as they are inherited from `StableDiffusionDepth2ImgPipeline`

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+            text_encoder,
+            tokenizer,
+            unet,
+            scheduler,
+            depth_estimator,
+            feature_extractor,
+        )
+
+        self.to(self._device)
+        torch.manual_seed(1)
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:186]: Why is this needed here?

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+    def get_timesteps(self, num_inference_steps, strength, device):
+        # get the original timestep using init_timestep
+        init_timestep = min(int(num_inference_steps * strength), num_inference_steps)
+
+        t_start = max(num_inference_steps - init_timestep, 0)
+        timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]
+        if hasattr(self.scheduler, "set_begin_index"):
+            self.scheduler.set_begin_index(t_start * self.scheduler.order)
+
+        return timesteps, num_inference_steps - t_start
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:198]: Same, it seems to be identical to https://github.com/huggingface/diffusers/blob/v0.29.2/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py#L464. We can remove it as it should be inherited from `StableDiffusionDepth2ImgPipeline`.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+            if torch.backends.mps.is_available():
+                autocast_ctx = contextlib.nullcontext()
+                logger.warning(
+                    "The DPT-Hybrid model uses batch-norm layers which are not compatible with fp16, but autocast is not yet supported on MPS."
+                )
+            else:
+                autocast_ctx = torch.autocast(device.type, dtype=dtype)
+
+            # --- HPU Patch --- #
+            with autocast_ctx:  # HPU Patch
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:287]:
```suggestion
            # --- HPU Patch --- #
            with autocast_ctx = torch.autocast(device.type, dtype=dtype):  # HPU Patch
```
We can remove the part about mps devices

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 12]
```diff
@@ -0,0 +1,653 @@
...
+    def do_classifier_free_guidance(self):
+        return self._guidance_scale > 1
+
+    @property
+    def cross_attention_kwargs(self):
+        return self._cross_attention_kwargs
+
+    @property
+    def num_timesteps(self):
+        return self._num_timesteps
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:341]: We can remove these, it will be inherited from `StableDiffusionDepth2ImgPipeline`

File: optimum/habana/diffusers/schedulers/scheduling_ddim.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pi314ever) [reply optimum/habana/diffusers/schedulers/scheduling_ddim.py:182]: I tested the DDIM scheduler with other models, it seems like it is only a problem with this model. I am uncertain of the root cause, but this is a sufficient patch in terms of output quality.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py (hunk 9 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:134]: Residual from debug testing, will remove

File: tests/test_diffusers.py (hunk 1 above)
Comment (pi314ever) [reply tests/test_diffusers.py:2242]: I added fidelity testing on the prompt. Is this sufficient?

**Comment by emascarenhas** [comment]
Please sync your PR with main/upstream and fix any merge conflicts. Thank you.

File: optimum/habana/diffusers/schedulers/scheduling_ddim.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pi314ever) [reply optimum/habana/diffusers/schedulers/scheduling_ddim.py:182]: I had to implement something similar in #1196 along with a patch to eliminate a bug with the scheduler that would cause unexpected `-inf` variance when `timestep = 1`.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 13]
```diff
@@ -0,0 +1,653 @@
...
+        image = 2.0 * image - 1.0
+        image = torch.from_numpy(image)
+    elif isinstance(image[0], torch.Tensor):
+        image = torch.cat(image, dim=0)
+    return image
+
+
+class GaudiStableDiffusionDepth2ImgPipeline(GaudiDiffusionPipeline, StableDiffusionDepth2ImgPipeline):
+    r"""
+    Adapted from: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py#L77
```
[the remark this answers is not in the document]
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:92]: No change

File: tests/test_diffusers.py (hunk 1 above)
Comment (regisss) [reply tests/test_diffusers.py:2246]: It looks good to me. However, the latency assert fails on my Gaudi2 instance:
```
assert 38.6319305896759 < (1.05 * 28.13371205329895)
```
Does it consistently succeed on yours?

**Comment by regisss** [comment]
@pi314ever I left a comment [here](https://github.com/huggingface/optimum-habana/pull/1175#discussion_r1744525574) because I don't get the right latency in `test_depth2img_pipeline_latency_bf16`.

File: tests/test_diffusers.py (hunk 1 above)
Comment (pi314ever) [reply tests/test_diffusers.py:2246]: It consistently passes for me after 12 attempts with no failures.

Comment (regisss) [reply tests/test_diffusers.py:2246]: Weird, I'll probably modify it later so that it passes on my instance.

**Review by regisss — APPROVED** [review]
